pkg/infoschema/infoschema_v2.go

```diff
@@ -1097,6 +1117,42 @@ func (is *infoschemaV2) SchemaTableInfos(ctx context.Context, schema ast.CIStr)
 	}
 
 	is.keepAlive()
+
+	// Fast path: all tables are in cache. Optimize for the few tables' scenario.
+	if is.tableCache.Under70PercentUsage() {
+		allInCache := true
+		db, ok := is.SchemaByName(schema)
+		if !ok || db == nil {
+			intest.Assert(false)
+			return nil, nil
+		}
+		is.IterateAllTableItemsByDB(db.ID, func(t TableItem) bool {
+			if t.DBName.L != schema.L {
+				return true
+			}
+			if !is.TableIsCached(t.TableID) {
+				allInCache = false
+				return false
+			}
+			return true
+		})
+		if allInCache {
+			tables := make([]*model.TableInfo, 0)
+			is.IterateAllTableItems(func(t TableItem) bool {
+				if t.DBName.L != schema.L {
+					return true
+				}
+				tbl, ok := is.TableByID(ctx, t.TableID)
+				intest.Assert(ok, "invalid table id", t.DBName.L)
+				tables = append(tables, tbl.Meta())
+				return true
```

Comment:
can we get the table via TableIsCached api? then we no need to get it again later if all in cache.